### PR TITLE
Add an Agent class

### PR DESCRIPTION
**Pull request by dlecocq**
The `Agent` class holds all user-agent-specific information, including the crawl-delay for the agent (if any), and the `Directive`s associated with that agent.

A path may match multiple directives, so the most-specific (i.e. longest) directive wins. For example, `/some/path/page.html` matches all of these rules:

```
Allow: /some/
Disallow: /some/path/
Allow: /*/page.html
```

As such, `/some/path/page.html` would be allowed because the `/*/page.html` rule is the longest.

Rather than finding all directives that match a query and having the highest priority one win, the `Directive`s in an `Agent` are sorted (lazily), so that we just have to traverse them until we find a match, and that is the winner. Future implementations might try to do something fancier, like a prefix trie, but I imagine that's far more trouble than it's worth.

This relates to https://github.com/seomoz/url-cpp/pull/28

@b4hand @arora-richa @tammybailey

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 CXX          ?= g++
 CXXOPTS      ?= -g -Wall -Werror -std=c++11 -Iinclude/ -Ideps/url-cpp/include
```
Comment (b4hand) [inline Makefile]: Typically, `-g` would be specified in just the `DEBUG_OPTS` and not in the `RELEASE_OPTS`. It appears you're passing it to both.

File: test/test-agent.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+TEST(AgentTest, DisallowNone)
+{
+    Rep::Agent agent = Rep::Agent().disallow("");
+    EXPECT_TRUE(agent.allowed("/anything"));
+}
+
+TEST(AgentTest, EscapedRule)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/a%3cd.html");
```
Comment (b4hand) [inline test/test-agent.cpp]: Question: Do you normalize to the unescaped rule and if so how does this interact with the longest rule match criteria. It seems like an escaped rule might need to overtake an unescaped rule due to length, but I'm not sure how that actually plays out and if this edge case is actually referenced anywhere.

File: test/test-agent.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    Rep::Agent agent = Rep::Agent().disallow("/a<d.html");
+    EXPECT_FALSE(agent.allowed("/a<d.html"));
+    EXPECT_FALSE(agent.allowed("/a%3cd.html"));
+}
+
+TEST(AgentTest, EscapedRuleWildcard)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/a%3c*");
+    EXPECT_FALSE(agent.allowed("/a<d.html"));
+    EXPECT_FALSE(agent.allowed("/a%3cd.html"));
```
Comment (b4hand) [inline test/test-agent.cpp]: A similar question applies to queries. Do you escape or unescape first before making the query and can that change which rule is matched due to length?

File: test/test-agent.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+TEST(AgentTest, UnescapedRuleWildcard)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/a<*");
+    EXPECT_FALSE(agent.allowed("/a<d.html"));
+    EXPECT_FALSE(agent.allowed("/a%3cd.html"));
+}
+
+TEST(AgentTest, AllowUrl)
```
Comment (b4hand) [inline test/test-agent.cpp]: This test name is not very good. It seems like you are testing params and/or queries, but that is not mentioned in the name at all.

Comment (tammybailey) [inline test/test-agent.cpp]: In other words, ur test name sux.

File: include/agent.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+
+namespace Rep
+{
+
+    class Agent
+    {
+    public:
+        /* The type for the delay. */
+        typedef float delay_t;
```
Comment (tammybailey) [inline include/agent.h]: Is there a particular reason you are creating this typedef?  It seems like you are adding an unnecessary level of obfuscation.

File: Makefile (hunk 1 above)
Comment (dlecocq) [inline Makefile]: kk

File: test/test-agent.cpp (hunk 2 above)
Comment (dlecocq) [inline test/test-agent.cpp]: The priority is taken after escaping. I don't know which is better, honestly. I think it's rare enough and also kind of a toss-up.

File: test/test-agent.cpp (hunk 3 above)
Comment (dlecocq) [inline test/test-agent.cpp]: All rules are escaped before being parsed, and all queries are escaped before checking.

File: test/test-agent.cpp (hunk 4 above)
Comment (dlecocq) [inline test/test-agent.cpp]: It's meant to test that it can accept a URL. Originally, I had meant to have two separate methods: `allowed` and `url_allowed`, the later of which accepts a full URL (rather than just a path).

However, in re-reading this, both have to go through the same pathway: `Url::Url(string).defrag().escape().fullpath()`, so I think I'll just collapse them into a single method.

I'll also rename this test to `AcceptsFullUrl`

File: include/agent.h (hunk 5 above)
Comment (dlecocq) [inline include/agent.h]: I tend to do this if I'm going to use that type in more than one place. If everyone thinks it's clunky, I don't feel strongly about this instance.

**Comment by dlecocq** [comment]
@b4hand @tammybailey Reupped

File: include/agent.h (hunk 5 above)
Comment (dlecocq) [inline include/agent.h]: Works for me. I can go to `double`. How do we feel about the typedef?
